File: 20.valid-parentheses.py
#
# @lc app=leetcode id=20 lang=python3
#
# [20] Valid Parentheses
#

# @lc code=start
class Solution:
    def isValid(self, s: str) -> bool:

        #use stack to simulate the process
        stack = []
        for c in s:
            if c in '([{':
                stack.append(c)
            else:
                if not stack:
                    return False
                if stack[-1] == '(' and c == ')':
                    stack.pop()
                elif stack[-1] == '[' and c == ']':
                    stack.pop()
                elif stack[-1] == '{' and c == '}':
                    stack.pop()
                else:
                    return False
        return not stack
# @lc code=end

File: 20.valid-parentheses.java
/*
 * @lc app=leetcode id=20 lang=java
 *
 * [20] Valid Parentheses
 */
import java.util.*;
// @lc code=start
class Solution {
    public boolean isValid(String s) {
        Deque<Character> stack = new ArrayDeque<>();
        for(char c : s.toCharArray()){
            if(c == '(' || c == '[' || c == '{'){
                stack.push(c);
            }
            else{
                if(!stack.isEmpty()){
                    if(c == ')' && stack.peek() == '(') stack.pop();
                    else if(c == ']' && stack.peek() == '[') stack.pop(); 
                    else if(c == '}' && stack.peek() == '{') stack.pop();
                    else stack.push(c);
                }else{
                    stack.push(c);
                }
            }
        }
        return stack.isEmpty();
    }
}
// @lc code=end

